components/brave_wallet/browser/meld_integration_response_parser.cc

```diff
@@ -0,0 +1,411 @@
+  //       "darkShort":
+  //       "https://images-serviceprovider.meld.io/BANXA/short_logo_dark.png",
+  //       "light":
+  //       "https://images-serviceprovider.meld.io/BANXA/logo_light.png",
+  //       "lightShort":
+  //       "https://images-serviceprovider.meld.io/BANXA/short_logo_light.png"
+  //    },
+  //    "name": "Banxa",
+  //    "serviceProvider": "BANXA",
+  //    "status": "LIVE",
+  //    "websiteUrl": "http://www.banxa.com"
+  // }
+
+  if (!json_value.is_list()) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is not a list";
+    return std::nullopt;
+  }
+
+  std::vector<mojom::MeldServiceProviderPtr> service_providers;
+  for (const auto& sp_item : json_value.GetList()) {
+    const auto service_provider_value =
+        meld_integration_responses::ServiceProvider::FromValue(sp_item);
+    if (!service_provider_value) {
+      LOG(ERROR)
+          << "Invalid response, could not parse JSON, JSON is not a dict";
+      return std::nullopt;
+    }
+
+    auto logos = ParseMeldLogos(service_provider_value->logos);
+    auto sp = mojom::MeldServiceProvider::New(
+        service_provider_value->name, service_provider_value->service_provider,
+        service_provider_value->status, service_provider_value->website_url,
+        service_provider_value->categories,
+        ParseOptionalMapOfStrings(service_provider_value->category_statuses),
+        std::move(logos).value_or(nullptr));
+
+    service_providers.emplace_back(std::move(sp));
+  }
+
+  if (service_providers.empty()) {
+    return std::nullopt;
+  }
+
+  return service_providers;
+}
+
+std::optional<std::vector<std::string>> ParseMeldErrorResponse(
+    const base::Value& json_value) {
+  // Parses results like this:
+  // {
+  //     "code": "BAD_REQUEST",
+  //     "message": "Bad request",
+  //     "errors": [
+  //         "[amount] Must be a decimal value greater than zero"
+  //     ],
+  //     "requestId": "eb6aaa76bd7103cf6c5b090610c31913",
+  //     "timestamp": "2022-01-19T20:32:30.784928Z"
+  // }
+
+  const auto meld_error_value =
+      meld_integration_responses::MeldError::FromValue(json_value);
+  if (!meld_error_value) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is not a dict";
+    return std::nullopt;
+  }
+
+  std::vector<std::string> errors;
+  if (meld_error_value->errors && !meld_error_value->errors->empty()) {
+    errors.assign(meld_error_value->errors->begin(),
+                  meld_error_value->errors->end());
+  }
+
+  if (meld_error_value->message.has_value() && errors.empty()) {
+    errors.emplace_back(*meld_error_value->message);
+  }
+
+  if (errors.empty()) {
+    return std::nullopt;
+  }
+
+  return errors;
+}
+
+std::tuple<std::optional<std::vector<mojom::MeldCryptoQuotePtr>>,
+           std::optional<std::string>>
+ParseCryptoQuotes(const base::Value& json_value) {
+  // Parses results like this:
```

Comment:
This allows both vector<MeldCryptoQuotePtr> and error be non-empty - is that possible?

Typical solution for "value or error" pattern is base::expected:
`base::expected<std::vector<mojom::MeldCryptoQuotePtr>, std::string>`

Reply:
fixed it